src/rust/engine/fs/store/src/local.rs

```diff
@@ -4,19 +4,345 @@ use super::{EntryType, ShrinkBehavior};
 
 use std::collections::{BinaryHeap, HashSet};
 use std::fmt::Debug;
-use std::io::{self, Read};
-use std::path::Path;
+use std::path::{Path, PathBuf};
 use std::sync::Arc;
 use std::time::{self, Duration, Instant};
 
+use async_trait::async_trait;
 use bytes::Bytes;
-use futures::future;
-use hashing::{Digest, Fingerprint, EMPTY_DIGEST};
+use futures::future::{self, join_all, try_join, try_join_all};
+use hashing::{async_copy_and_hash, async_verified_copy, Digest, Fingerprint, EMPTY_DIGEST};
 use lmdb::Error::NotFound;
 use lmdb::{self, Cursor, Transaction};
 use sharded_lmdb::{ShardedLmdb, VersionedFingerprint};
+use std::os::unix::fs::PermissionsExt;
+use tempfile::NamedTempFile;
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
 use workunit_store::ObservationMetric;
 
+/// How big a file must be to be stored as a file on disk.
+// NB: These numbers were chosen after micro-benchmarking the code on one machine at the time of
+// writing. They were chosen using a rough equation from the microbenchmarks that are optimized
+// for somewhere between 2 and 3 uses of the corresponding entry to "break even".
+const LARGE_FILE_SIZE_LIMIT: usize = 512 * 1024;
+
+#[derive(Debug, Clone)]
+pub struct TempImmutableLargeFile {
+  tmp_path: PathBuf,
+  final_path: PathBuf,
+}
+
+impl TempImmutableLargeFile {
+  pub async fn open(&self) -> tokio::io::Result<tokio::fs::File> {
+    tokio::fs::File::create(self.tmp_path.clone()).await
+  }
+
+  pub async fn persist(&self) -> Result<(), String> {
+    tokio::fs::rename(self.tmp_path.clone(), self.final_path.clone())
+      .await
+      .map_err(|e| format!("Error while renaming: {e}."))?;
+    tokio::fs::set_permissions(&self.final_path, std::fs::Permissions::from_mode(0o555))
+      .await
+      .map_err(|e| e.to_string())?;
+    Ok(())
+  }
+}
+
+/// Trait for the underlying storage, which is either a ShardedLMDB or a ShardedFS.
+#[async_trait]
+trait UnderlyingByteStore {
```

Comment:
So I understand, is this trait used anywhere? It seems like it may be defining a common interface, but there's never a `<T: UnderlyingByteStore>` generic or `Arc<dyn UnderlyingByteStore>` or similar?

Reply:
Yeah I defined the trait thinking in the future the consumer of the bytestore API would be getting an implementation so the caller can also decide how to store things.

But I think that's probably best handled via params. 

We can probably nix this.

Bear with me I'm still learning 🙂

Reply:
I like the trait, even if it isn't actually being used. Although it has a large number of methods (which makes it more challenging to "stack" correctly), it's possible that it could end up being used fully generically eventually, similar to `CommandRunner`.